### PR TITLE
Bug 2061451: Fix UI crash when D/DC name not available

### PR DESCRIPTION
**Pull request by divyanshiGupta**
**Fixes:** https://issues.redhat.com/browse/OCPBUGSM-41417

**Screenshots:**
Screenshot 2022-03-21 at 4 52 39 PM
Screenshot 2022-03-21 at 4 54 08 PM
Screenshot 2022-03-21 at 6 58 43 PM

/kind bug

### CODE REVIEW

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/src/components/edit-deployment/EditDeploymentPage.tsx [hunk 1]
```diff
@@ -17,7 +17,7 @@ const EditDeploymentPage: React.FC<EditDeploymentPageProps> = ({ match, location
   const kind = queryParams.get('kind');
   const name = queryParams.get('name');
   const heading = t('devconsole~Edit {{kind}}', { kind });
-  const label = t('devconsole~Edit {{kind}} form', { kind });
+  const label = t('devconsole~{{kind}}', { kind });
```
[the remark this answers is not in the document]
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/edit-deployment/EditDeploymentPage.tsx]: Done.

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/src/components/edit-deployment/EditDeploymentPage.tsx [hunk 2]
```diff
@@ -16,8 +16,15 @@ const EditDeploymentPage: React.FC<EditDeploymentPageProps> = ({ match, location
   const queryParams = new URLSearchParams(location.search);
   const kind = queryParams.get('kind');
   const name = queryParams.get('name');
+  const kindModel = modelFor(kind) ?? {};
```
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/edit-deployment/EditDeploymentPage.tsx:19]: Now it crashes if the kind parameter is missing.

Error: `Cannot read properties of null (reading 'split')`

**Review by jerolimov — CHANGES_REQUESTED** [review]
I know that this is unexpected. But this bug is all about the unexpected URL changes so we should not decrease our stability.

Can you please also add some `console.errors` to `useK8sWatchResource` and `useK8sWatchResources` if `isList` and `name` are falsy. The bug report was also that we harden these hooks.

It could happen on any page that we miss (or the user doesn't pass) a name when we expect a single item. And instead of changing or breaking any code now I would suggest just a "warning" (console.error) which is visible to our developers, but this will not break anything.

**Comment by divyanshiGupta** [comment]
Confirmed with Sam that getting all resources when isList is false and resource name is not provided is a bug. Handled it similar to how its being handled in dynamic-plugin-sdk https://github.com/openshift/dynamic-plugin-sdk/blob/95a6baecc418251575a5f78ad9b7d9cb97626a60/packages/lib-utils/src/k8s/hooks/k8s-watcher.ts#L79-L113 alongside fixing the issue wherein when name is not provided and isList is true it does not return the resource list and  another issue being that not all resources are namespaced and hence adding a namespace should be optional. 

@jerolimov PTAL.

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/useK8sWatchResource.ts [hunk 3]
```diff
@@ -64,8 +64,10 @@ export const useK8sWatchResource: UseK8sWatchResource = (initResource) => {
     }
     if (!resourceK8s) {
       const data = resource?.isList ? [] : {};
-      return modelsLoaded && !k8sModel
-        ? [data, true, new NoModelError()]
+      return modelsLoaded
+        ? !k8sModel
+          ? [data, true, new NoModelError()]
+          : [data, true, undefined]
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/useK8sWatchResource.ts:70]: Any reason you decided the suppress the error?

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts [hunk 4]
```diff
@@ -71,7 +71,7 @@ export const getReduxData = (immutableData, resource: WatchK8sResource) => {
 };
 
 export const getIDAndDispatch: GetIDAndDispatch<SDKStoreState> = (resource, k8sModel, cluster) => {
-  if (!k8sModel || !resource) {
+  if (!k8sModel || !resource || (!resource.isList && !resource.name)) {
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts:74]: I think we should look at updating the type too, no? If this is not a valid state, we should make TS step in the way and say the call is bad. I think in your case however, the dynamic nature of how you get `name`, TS might not help out here.

In principle, however, I think we should update our types. Thoughts?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts (hunk 4 above)
Comment (divyanshiGupta) [reply frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts:74]: So `isList` and `name` are optional fields and its reflected in the `WatchK8sResource` as well. Can you please explain which type you are talking about and how should we update it?

Comment (andrewballantyne) [reply frontend/packages/console-dynamic-plugin-sdk/src/utils/k8s/hooks/k8s-watcher.ts:74]: There is no relationship between the two fields (`isList` & `name`) in Typescript but... you're stating here that `if I am not isList and I do not have a name, that's a bad state`. We should try to showcase that in the Types so we can properly establish typings to match logic and get away from pure runtime checks when we have TypeScript.

### But, as I was testing it out, to no surprise to anyone I'm sure, OpenShift Console has massively poor types and this following solution breaks all kinds of code from a TS point of view. I'll log a bug to address this and you can ignore this comment.

But for your own learning experience, this is what I was thinking...

---

`WatchK8sResource` got a facelift in [openshift/dynamic-plugin-sdk](https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/lib-utils/src/k8s/hooks/watch-resource-types.ts#L25-L39) to make the API more idiomatic (ie optional everything is a poor type). We could follow that logic and expand it to `!isList` MUST have `name`, imo.

Something like:
```
export type WatchK8sResource = EitherNotBoth<
  { kind: K8sResourceKindReference },
  { groupVersionKind: K8sGroupVersionKind }
> &
  EitherNotBoth<{ isList: true }, { name: string }> &
  Partial<{
    namespace: string;
    selector: Selector;
    namespaced: boolean;
    limit: number;
    fieldSelector: string;
    optional: boolean;
    partialMetadata: boolean;
    cluster: string;
  }>;
```
`EitherNotBoth` Generic type comes [from here](https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/common/src/types/common.ts#L27-L90).

**Review by jerolimov — APPROVED** [review]
I personally would deprecate this "behaviour" first with a `console.error`, but I'm also fine with this solution. Thanks Divyanshi.

/lgtm
/approve

**Comment by jerolimov** [comment]
Looks like this change breaks the admin perspective e2e tests. Could also not open the Create Route page there when running this locally.

It fails if you open Admin perspective > Networking > Routes > Create Route

See also https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/11208/pull-ci-openshift-console-master-e2e-gcp-console/1514590052666052608/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/cypress/screenshots/crud/resource-crud.spec.ts/

/approve cancel
/lgtm cancel

**Review by jerolimov — CHANGES_REQUESTED** [review]
See above.

**Pull request closed without merging**